lib/routes/account.js

```diff
@@ -973,78 +977,89 @@ module.exports = (log, db, mailer, Password, config, customs, signinUtils, push)
-                // Notify the devices that the account has changed.
-                request.app.devices.then(devices =>
-                  push.notifyPasswordReset(accountResetToken.uid, devices)
-                )
-
-                return db.account(accountResetToken.uid)
-                  .then((accountData) => {
-                    account = accountData
-                  })
-              }
-            )
-            .then(
-              function () {
-                return P.all([
-                  request.emitMetricsEvent('account.reset', {
-                    uid: account.uid
-                  }),
-                  log.notifyAttachedServices('reset', request, {
-                    uid: account.uid,
-                    iss: config.domain,
-                    generation: account.verifierSetAt
-                  }),
-                  customs.reset(account.email),
-                  password.unwrap(account.wrapWrapKb)
-                ])
-              }
-            )
-            .then(
-              function (results) {
-                wrapKb = results[3]
+            .then((verifyHashData) => {
+              verifyHash = verifyHashData
+
+              // If this is an account recovery, reuse the specified wrapKb instead of creating
+              // a random one. The wrapKb specified should contain the original kB of the account.
+              if (recoveryKeyId) {
+                return password.wrap(wrapKb)
               }
-            )
+
+              return random.hex(32)
+            })
+            .then((result) => {
+              wrapWrapKb = result
+              return db.resetAccount(
+                accountResetToken,
+                {
+                  authSalt,
+                  verifyHash,
+                  wrapWrapKb,
+                  verifierVersion: password.version
+                }
+              )
+            })
+            .then(() => {
+              // Delete all passwordChangeTokens, passwordForgotTokens and
+              // accountResetTokens associated with this uid
+              return db.resetAccountTokens(accountResetToken.uid)
+            })
+            .then(() => {
+              // Notify the devices that the account has changed.
+              request.app.devices.then(devices => push.notifyPasswordReset(accountResetToken.uid, devices))
+
+              return db.account(accountResetToken.uid)
+                .then((accountData) => account = accountData)
+            })
+            .then(() => {
+              return P.all([
+                request.emitMetricsEvent('account.reset', {
+                  uid: account.uid
+                }),
+                log.notifyAttachedServices('reset', request, {
+                  uid: account.uid,
+                  iss: config.domain,
+                  generation: account.verifierSetAt
+                }),
+                customs.reset(account.email),
+                password.unwrap(wrapWrapKb)
```

Comment:
Calling `password.unwrap(wrapWrapKb)` does unnecessary work in the case where `wrapKb` was provided.  I wonder if we can clean up the code a bit here by moving it up to the point where we generate a random `wrapWrapKb`, and only doing it if necessary.  Something like:

```
if (recoveryKeyId) {
  // We have the previous kB, just re-wrap it with the new password.
  return password.wrap(wrapKb).then(result => {
    wrapWrapKb = result
  })
} else {
  // We need to regenerate kB and wrap it with the new password.
  return random.hex(32).then(result => {
    wrapWrapKb = result
    return password.unwrap(wrapWrapKb).then(result => {
      wrapKb = result
    }
  }
}
```

This would keep all the kb-related things together in one place in the code, which might be helpful for future readers.